components/eamxx/src/physics/mam/eamxx_mam_aci_process_interface.cpp

```diff
@@ -0,0 +1,1299 @@
+    add_field<Updated>(gas_mmr_field_name, scalar3d_layout_mid, q_unit,
+                       grid_name, "tracers");
+  }  // end for loop num gases
+
+  // ------------------------------------------------------------------------
+  // Output from ice nucleation process
+  // ------------------------------------------------------------------------
+
+  // number of activated aerosol for ice nucleation[#/kg]
+  add_field<Computed>("ni_activated", scalar3d_layout_mid, n_unit, grid_name);
+
+  // ------------------------------------------------------------------------
+  // Output from droplet activation process (dropmixnuc)
+  // ------------------------------------------------------------------------
+
+  // tendency in droplet number mixing ratio [#/kg/s]
+  add_field<Computed>("nc_nuceat_tend", scalar3d_layout_mid, n_unit / s,
+                      grid_name);
+
+  // ------------------------------------------------------------------------
+  // Output from hetrozenous freezing
+  // ------------------------------------------------------------------------
+
+  const auto cm = m / 100;
+
+  // units of number mixing ratios of tracers
+  auto frz_unit = 1 / (cm * cm * cm * s);
+  n_unit.set_string("1(cm^-3 s^-1)");
+  // heterogeneous freezing by immersion nucleation [cm^-3 s^-1]
+  add_field<Computed>("hetfrz_immersion_nucleation_tend", scalar3d_layout_mid,
+                      frz_unit, grid_name);
+
+  // heterogeneous freezing by contact nucleation [cm^-3 s^-1]
+  add_field<Computed>("hetfrz_contact_nucleation_tend", scalar3d_layout_mid,
+                      frz_unit, grid_name);
+
+  // heterogeneous freezing by deposition nucleation [cm^-3 s^-1]
+  add_field<Computed>("hetfrz_depostion_nucleation_tend", scalar3d_layout_mid,
+                      frz_unit, grid_name);
+}  // function set_grids ends
+
+// ================================================================
+//  INIT_BUFFERS
+// ================================================================
+
+void MAMAci::init_buffers(const ATMBufferManager &buffer_manager) {
+  EKAT_REQUIRE_MSG(
+      buffer_manager.allocated_bytes() >= requested_buffer_size_in_bytes(),
+      "Error! Insufficient buffer size.\n");
+
+  size_t used_mem =
+      mam_coupling::init_buffer(buffer_manager, ncol_, nlev_, buffer_);
+  EKAT_REQUIRE_MSG(
+      used_mem == requested_buffer_size_in_bytes(),
+      "Error! Used memory != requested memory for MAMMicrophysics.");
+}  // function init_buffers ends
+
+// ================================================================
+//  INITIALIZE_IMPL
+// ================================================================
+void MAMAci::initialize_impl(const RunType run_type) {
+  // ------------------------------------------------------------------------
+  // ## Runtime options
+  // ------------------------------------------------------------------------
+
+  wsubmin_ = m_params.get<double>("wsubmin");
+  top_lev_ = m_params.get<int>("top_level_mam4xx");
```

Comment:
How is `top_lev` in E3SM v2 obtained? is it also passed as input parameter?

Reply:
We will include a code to compute it. It varies with different vertical resolutions.